raw-spans-grouper/raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansProcessor.java

```diff
@@ -196,7 +242,108 @@ public void process(Record<TraceIdentity, RawSpan> record) {
+            currentTimeMs, firstEntry ? currentTimeMs : traceState.getTraceStartTimestamp());
+    StructuredTrace trace =
+        StructuredTraceBuilder.buildStructuredTraceFromRawSpans(
+            List.of(value), traceId, tenantId, timestamps);
+    context.forward(new Record<>(key, trace, currentTimeMs), OUTPUT_TOPIC_PRODUCER);
+  }
+
+  // put the peer service identity and corresponding service name in state store
+  private void handleClientSpan(String tenantId, Event event, String environment) {
+    Optional<String> maybeHostAddr = HttpSemanticConventionUtils.getHostIpAddress(event);
+    Optional<String> maybePeerAddr = HttpSemanticConventionUtils.getPeerIpAddress(event);
+    Optional<String> maybePeerPort = HttpSemanticConventionUtils.getPeerPort(event);
+    String serviceName = event.getServiceName();
+    PeerIdentity peerIdentity =
+        PeerIdentity.newBuilder()
+            .setIpIdentity(
+                IpIdentity.newBuilder()
+                    .setTenantId(tenantId)
+                    .setEnvironment(environment)
+                    .setHostAddr(maybeHostAddr.orElse(null))
+                    .setPeerAddr(maybePeerAddr.orElse(null))
+                    .setPeerPort(maybePeerPort.orElse(null))
+                    .build())
+            .build();
+    if (PeerIdentityValidator.isValid(peerIdentity)) {
+      this.peerIdentityToSpanMetadataStateStore.put(
+          peerIdentity, SpanMetadata.newBuilder().setServiceName(serviceName).build());
+    }
+  }
+
+  // get the service name for that peer service identity and correlate the current span
+  private void handleServerSpan(String tenantId, Event event, String environment) {
+    Optional<String> maybePeerAddr = HttpSemanticConventionUtils.getPeerIpAddress(event);
+    Optional<String> maybeHostAddr = HttpSemanticConventionUtils.getHostIpAddress(event);
+    Optional<String> maybeHostPort = HttpSemanticConventionUtils.getHostPort(event);
+    PeerIdentity peerIdentity =
+        PeerIdentity.newBuilder()
+            .setIpIdentity(
+                IpIdentity.newBuilder()
+                    .setTenantId(tenantId)
+                    .setEnvironment(environment)
+                    .setHostAddr(maybePeerAddr.orElse(null))
+                    .setPeerAddr(maybeHostAddr.orElse(null))
+                    .setPeerPort(maybeHostPort.orElse(null))
+                    .build())
+            .build();
+    if (PeerIdentityValidator.isValid(peerIdentity)) {
+      SpanMetadata spanMetadata = this.peerIdentityToSpanMetadataStateStore.get(peerIdentity);
+      if (Objects.nonNull(spanMetadata)) {
+        // Instead of updating the span, adding a debug log to print the service correlation
+        logger.debug(
+            "Adding {} as: {} in spanId: {} with service name: {}",
+            PEER_SERVICE_NAME,
+            spanMetadata.getServiceName(),
+            HexUtils.getHex(event.getEventId()),
+            event.getServiceName());
+
+        event
```

Comment:
i think we should add this in enrichedAttributes not attribute map. @skjindal93 I believe enriched attributes are for anything that traceable resolves and add to the span

Reply:
adding to enriched attributes since we changed the attribute name to `PEER_SERVICE_NAME`